### DISCUSSION TITLE
Generated single page PDF is huge

### DISCUSSION THREAD

**Discussion by Vafilor** [py-pdf/pypdf#3012] [Q&A] [answered]
I am trying to split a PDF into pages. The original PDF is 71 M, 898 pages. The resulting first page as a PDF is 69 M.
This is the PDF: https://cdn-resources.ableton.com/resources/pdfs/live-manual/12/2024-11-13/live12-manual-en.pdf

## Environment

Which environment were you using when you encountered the problem?

```bash
$ python -m platform

macOS-15.1.1-arm64-arm-64bit-Mach-O

$ python -c "import pypdf;print(pypdf._debug_versions)"

pypdf==5.1.0, crypt_provider=('local_crypt_fallback', '0.0.0'), PIL=none
```

## Code + PDF

PDF: https://cdn-resources.ableton.com/resources/pdfs/live-manual/12/2024-11-13/live12-manual-en.pdf

```python
from pypdf import PdfReader, PdfWriter
from pathlib import Path

# Assumes the downloaded PDF is in the same directory as this script
# Change if needed
reader = PdfReader("./live12-manual-en.pdf")
writer = PdfWriter()

writer.add_page(reader.pages[0])

# Doesn't seem to do anything
# for page in writer.pages:
# 	page.compress_content_streams()

output_path = Path("./page_1.pdf")
writer.write(output_path)

size = output_path.stat().st_size / 1024 / 1024
print(f"Page 1 size: {size} M")

```

Output
`Page 1 size: 69.28027153015137 M`

## Traceback

N/A

**Comment by pubpub-zz** [comment, marked as answer]
this PDF is an example of  files where the ressources shared between all the pages.
![image](https://github.com/user-attachments/assets/b4317d92-9fb7-447f-9740-30968706b401)
the full /Resources is shared between all pages, so cleaning should be done carefully.
considering a case where many pages could be added, this is a (draft) proposal for clean up;
```
from pypdf import PdfReader, PdfWriter
from pathlib import Path
from io import BytesIO

# Assumes the downloaded PDF is in the same directory as this script
# Change if needed
reader = PdfReader("./live12-manual-en.pdf")
writer = PdfWriter()

writer.add_page(reader.pages[0])
writer.add_page(reader.pages[26])  # to get pages with some images
writer.add_page(reader.pages[28]) # to get pages with some images

used_obj = []
for p in writer.pages:
   used_obj.extend([ope for (ope,op) in p.get_contents().operations if ope == b"Do"])

# only applies as resources are shared 
xobjects = writer.pages[0]["/Resources"]["/XObject"].get_object()
for k in tuple(xobjects.keys()):
   if k not in used_obj:
      del xobjects[k]

output_path = Path("./page_1.pdf")
# getting through an intermediate is required to get ride of unused objects
b = BytesIO()
writer.write(b)
PdfWriter(clone_from:b).write(output_path)

size = output_path.stat().st_size / 1024 / 1024
print(f"Page 1 size: {size} M")
```

**Reply by stefan6419846** [reply]
Apparently, you have been a bit faster than me with typing an answer ;) I will leave my answer for further clarification nevertheless as it documents how to spot this from the actual PDF source as well.

**Reply by Vafilor** [reply]
Thanks for the quick reply! 

This trims down the first page of the PDF a ton (0.037 M vs 69 M). It looks like it does strip out the `Ableton` logo on the first page though, but I think I can work with what you have above to fix that.

**Comment by stefan6419846** [comment]
If you have a look at the internal structure of the generated PDF file, you will see that even the one-page file contains all images:

![Structure viewer](https://github.com/user-attachments/assets/aaa6f652-bd45-48aa-8edd-14bd433abed3)

This is not directly obvious from looking at the content of the original PDF file directly, but after decoding the content streams (`pdftk live12-manual-en.pdf output out.pdf uncompress`), we will see an object which looks like this holding lots of references to other objects:

```
902 0 obj

<<
/ExtGState 
<<
/a1.0 
<<
/ca 1
>>
/a1 
<<
/ca 1
>>
>>
/Shading 
<<
>>
/XObject 
<<
/id3e6618495497251be9c6798d6e1705c1 903 0 R
/i9e41ecb724394f9ef756d89129a2b5411 904 0 R
/i11309e6cccbd061fbf814195ebe4f4c71 905 0 R
/i7f233022b6e21bc1415bc384d4e662961 906 0 R
/i9a16cb807637a8b6d9ef9d0c7ca8aded1 907 0 R
/i3ab73536bea5b843d8da75777c1ced721 908 0 R
/if6a45e5ceaa1799884c0ac3f703c32c31 909 0 R
...
/id7bb5adc50d015c62f407a68840882371 1860 0 R
/i77aa06585640dfcfbedac46986e494e01 1861 0 R
>>
/Pattern 
<<
>>
/Font 1862 0 R
>>
endobj
```

If you search for further occurrences of this xobject 902 (the number might differ due to the work done by *pdftk* for the decompression), you will discover that it is referenced by basically every page:

```
2 0 obj

<<
/pdftk_PageNum 1
/Contents 1863 0 R
/BleedBox [0 0 595.275591 841.889764]
/Type /Page
/Resources 902 0 R
/Parent 1 0 R
/Annots [1864 0 R]
/MediaBox [0 0 595.275591 841.889764]
/TrimBox [0 0 595.275591 841.889764]
>>
endobj
```

```
3 0 obj

<<
/pdftk_PageNum 2
/Contents 1865 0 R
/BleedBox [0 0 595.275591 841.889764]
/Type /Page
/Resources 902 0 R
/Parent 1 0 R
/Annots [1866 0 R 1867 0 R 1868 0 R 1869 0 R 1870 0 R 1871 0 R 1872 0 R 1873 0 R 1874 0 R 1875 0 R 1876 0 R 1877 0 R 1878 0 R 1879 0 R 1880 0 R 1881 0 R 1882 0 R 1883 0 R 1884 0 R 1885 0 R 1886 0 R 1887 0 R 1888 0 R 1889 0 R 1890 0 R 1891 0 R 1892 0 R 1893 0 R 1894 0 R 1895 0 R 1896 0 R 1897 0 R 1898 0 R 1899 0 R 1900 0 R 1901 0 R 1902 0 R 1903 0 R 1904 0 R 1905 0 R 1906 0 R 1907 0 R 1908 0 R 1909 0 R 1910 0 R 1911 0 R 1912 0 R 1913 0 R 1914 0 R 1915 0 R 1916 0 R 1917 0 R 1918 0 R 1919 0 R 1920 0 R 1921 0 R 1922 0 R 1923 0 R 1924 0 R 1925 0 R 1926 0 R 1927 0 R 1928 0 R 1929 0 R 1930 0 R 1931 0 R 1932 0 R 1933 0 R 1934 0 R 1935 0 R 1936 0 R 1937 0 R 1938 0 R 1939 0 R 1940 0 R 1941 0 R 1942 0 R 1943 0 R 1944 0 R 1945 0 R 1946 0 R 1947 0 R 1948 0 R 1949 0 R 1950 0 R 1951 0 R 1952 0 R 1953 0 R 1954 0 R 1955 0 R 1956 0 R 1957 0 R 1958 0 R 1959 0 R 1960 0 R 1961 0 R 1962 0 R 1963 0 R 1964 0 R 1965 0 R 1966 0 R 1967 0 R]
/MediaBox [0 0 595.275591 841.889764]
/TrimBox [0 0 595.275591 841.889764]
>>
endobj
```

It basically is the fault of the original creator of the PDF file that there is no clean separation and you end up with nearly the same size as the original file for each page.

**Reply by Vafilor** [reply]
Thank you for the instructions, that really helps!

I'll see if I can't strip out the duplicates to get smaller pages.

**Comment by shartzog** [comment]
@Vafilor, if you're still struggling with this you might find some inspiration [here](https://github.com/py-pdf/pypdf/discussions/2767#discussioncomment-10168988). I've been using similar logic in my pipelines for quite a while with no adverse effects (like the loss of your logo image).

**Reply by ZupoLlask** [reply]
Interesting project you got there, @shartzog... 😉

**Reply by shartzog** [reply]
Still using more or less the same method in my codebase to this day... It ain't broke, so I don't fix it... 😆
